File: Python/Linked-List/SinglyLinkedList.py
class Node:
    # constructor
    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next

class LinkedList:
    def __init__(self):
        self.head = None

    # insertion method
    def insertNode(self, ele):
        newNode = Node(ele)
        if(self.head):
            current = self.head
            while(current.next):
                current = current.next
            current.next = newNode
        else:
            self.head = newNode

    # deletion method
    def deleteNode(self, ele):
        p1 = self.head
        # if head node is the element to be deleted
        if p1 is not None:
            if p1.data == ele:
                self.head = p1.next
                p1 = None
                print("!!! ELEMENT DELETED !!!")
                return
        
        # search ele in list
        while p1 is not None:
            if p1.data == ele:
                break
            p2 = p1
            p1 = p1.next
        
        # if element not present in list
        if p1 == None:
            print("*** ELEMENT NOT FOUND ***")
            return
        
        # delete the node
        p2.next = p1.next
        p1 = None

    # search method
    def searchNode(self, ele):
        p1 = self.head
        # head element to be searched
        if p1.data == ele:
            print("!!! ELEMENT FOUND !!!")
            return

        while p1 is not None:
            if p1.data == ele:
                print("!!! ELEMENT FOUND !!!")
                return
            p1 = p1.next
            
        print("*** ELEMENT NOT FOUND ***")
        return

    # print method
    def printList(self):
        current = self.head
        while(current):
            print(current.data, end=" ")
            current = current.next


def main():
    List = LinkedList()
    ch = 'y'
    # input code
    while ch == 'y':
        print("Choose an operation :\n1. Insertion\n2. Deletion\n3. Search\n4. Print\n\tMAKE YOUR CHOICE : ", end="")
        c = int(input())
        # insertion
        if c == 1:
            print("\nINSERTION\nEnter List Size : ", end="")
            n = int(input())
            print("\nEnter List Elements : ", end="")
            eleList = [int(i) for i in input().split()][:n]
            for i in eleList:
                List.insertNode(i)

        # deletion
        elif c == 2:
            print("DELETION\nEnter element for deletion : ", end="")
            ele = int(input())
            List.deleteNode(ele)
        
        # search
        elif c == 3:
            print("SEARCHING\nEnter element to be searched : ", end="")
            ele = int(input())
            List.searchNode(ele)

        # print
        elif c == 4:
            print("List Elements : ", end="")
            List.printList()
        
        else:
            print("\n\t*** INVALID INPUT ***")
        
        ch = input("\n\tTry Again ? (y/n) : ")

main()

File: C++/Linked-List/SinglyLinkedList.cpp
#include <iostream>
#include <stdio.h>
using namespace std;
class Node {
public:
    int data;
    Node *next;
    Node() {
        data = 0;
        next = NULL;
    }
    Node(int data) {
        this->data = data;
        this->next = NULL;
    }
};

class LinkedList {
    Node *head;

public:
    LinkedList() { head = NULL; }
    void insertNode(int);
    void deleteNode(int);
    void searchNode(int);
    void printList();
};

// insertion method
void LinkedList::insertNode(int x) {
    Node *newNode = new Node(x);
    if (!head)
    {
        head = newNode;
        return;
    } // if empty list
    Node *temp = head;
    while (temp->next)
        temp = temp->next;
    temp->next = newNode;
}

// deletion method
void LinkedList::deleteNode(int pos) {
    Node *p1 = head, *p2 = NULL;
    int length = 0;
    if (!head) { cout << "EMPTY LIST" << endl; return; }
    // list length
    while (p1) { p1 = p1->next; ++length; }

    // check if postion less than list length
    if (length < pos) { cout << "Index Out Of Range" << endl; return; }

    p1 = head;
    // if haed to be deleted
    if (pos == 1) {
        head = head->next;
        delete p1;
        cout << "\n\t!!! ELEMENT DELETED !!!";
        return;
    }
    while (pos-- > 1) { p2 = p1; p1 = p1->next; }
    p2->next = p1->next;
    delete p1;
    cout << "\n\t!!! ELEMENT DELETED !!!";
}

// search method
void LinkedList::searchNode(int x) {
    Node *p = head;
    int flag = 0;
    while (p)
    {
        p = p->next;
        if (p->data == x)
        {
            flag = 1;
            break;
        }
    }
    if (flag) cout << "\n!!! ELEMENT FOUND !!!";
    else cout << "\n*** ELEMENT NOT FOUND ***";
}

// print method
void LinkedList::printList() {
    Node *p = head;
    if (head == NULL) { cout << "Empty List" << endl; return; }
    while (p) { cout << p->data << " "; p = p->next; }
}

int main() {
    LinkedList list;
    int n, ele, c;
    char ch = 'y';

    // Input code
    while (ch == 'y') {
        cout << "\nChoose an operation :\n1. Insertion\n2. Deletion\n3. Search\n4. Print";
        cout << "\n\tMAKE YOU CHOICE : ";
        cin >> c;
        switch (c) {
        case 1: // Insertion
            cout << "\nINSERTION\nEnter List Size : ";
            cin >> n;
            cout << "\nEnter List Elements : ";
            for (int i = 0; i < n; i++) { cin >> ele; list.insertNode(ele); }
            break;

        case 2: // Deletion
            cout << "\nDELETION\nEnter the position for deletion : ";
            cin >> ele;
            list.deleteNode(ele);
            break;

        case 3: // Searching
            cout << "\nSEARCHING\nEnter the element to be searched : ";
            cin >> ele;
            list.searchNode(ele);
            break;

        case 4: // Print
            cout << "\nList Elements : ";
            list.printList();
            break;

        default:
            cout << "\n\t*** INVALID INPUT ***";
            break;
        }
        cout << "\n\tTry Again ? (y/n) : ";
        cin >> ch;
    }
    return 0;
}